presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -183,13 +183,23 @@ public ActualProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<Ac
         public ActualProperties visitAssignUniqueId(AssignUniqueId node, List<ActualProperties> inputProperties)
         {
             ActualProperties properties = Iterables.getOnlyElement(inputProperties);
+
+            ImmutableList.Builder<LocalProperty<Symbol>> newLocalProperties = ImmutableList.builder();
+            newLocalProperties.addAll(properties.getLocalProperties());
+            newLocalProperties.add(new GroupingProperty<>(ImmutableList.of(node.getIdColumn())));
```

Comment:
I think those should be new top-level (stronger) properties before previous local properties.

Reply:
@sopel39 Here is an example to show that adding new local properties to the end of the list is better.

Let's say data is sorted on `a`. Current logic generates `S(a), G(unique), C(a), C(b)` properties, which say:

- Data is sorted in `a`
- Within each group of `a`, data is grouped on `unique`
- Within each group of (`a`, `unique`), `a` is constant (this is redundant)
- Within each group of (`a`, `unique`), `b` is constant.

These properties match `S(a)` requirement as well as `G(unique)` and `G(unique, a, b)`.

If we change these to `G(unique), C(a), C(b), S(a)`, the new properties will no longer match `S(a)` requirement.

Reply:
@sopel39 

> But previous top-level local properties become some nested properties in local properties hierarchy. Is that OK? It's not longer a leading constant (as ActualProperties#constants seem to be)

I don't undestand this. Could you give an example?

Reply:
> I don't undestand this. Could you give an example?

If you have a leading constant property `C1` and you prepend it with `Grouping(X)` then `C1` is no longer a leading property, but a property which is below `Grouping(X)`.

Reply:
> If we change these to G(unique), C(a), C(b), S(a), the new properties will no longer match S(a) requirement.

True, but on a flip side if you append `G(unique)` you won't match grouping on `unique`, no?

Reply:
> These properties match S(a) requirement as well as G(unique) and G(unique, a, b).

I think this means that `grouping` as a local property should always be last, because it's satisfied by other properties. Sounds good

Reply:
@sopel39 

> True, but on a flip side if you append G(unique) you won't match grouping on unique, no?

You are correct. I got it wrong in my comment. We get `S(a)` and `G(unique, a)` and `G(unique, a, b)`, but not `G(unique)`. I edited that comment.

I think there is no perfect solution. Having `G(unique), C(a), C(b)` appended allows to preserve the original properties and satisfy streaming aggregation requirements. Prepending same properties makes us lose any original properties (everything becomes scoped to `within unique`).

I'm inclined to keep `G(unique), C(a), C(b)` appended until we have richer support for properties unless you feel strongly otherwise.